File: src/main/api/object.ts
import { generateRefferer, generateFieldsBasedOnModel, generateSiteId } from './utils'

interface Config {
  baseUrl: string
  userAgent: string
  test?: boolean
}

export class ObjectManager {
  private config
  constructor(config: Config) {
    this.config = config
  }
  async getObject(fields: string[] = [], model: string, slug_id: string): Promise<string> {
    if (fields.length === 0) {
      fields = generateFieldsBasedOnModel(model)
    }
    const params = fields.map((field) => `fields[]=${encodeURIComponent(field)}`).join('&')
    try {
      const response = await fetch(
        `https://${this.config.baseUrl}/api/${model}/${slug_id}?${params.toString()}`,
        {
          credentials: 'include',
          headers: {
            'User-Agent': this.config.userAgent,
            Accept: '*/*',
            'Accept-Language': 'ru-RU,en-US,en',
            'Site-Id': generateSiteId(model),
            'Content-Type': 'application/json',
            'Client-Time-Zone': Intl.DateTimeFormat().resolvedOptions().timeZone,
            'Sec-Fetch-Dest': 'empty',
            'Sec-Fetch-Mode': 'cors',
            'Sec-Fetch-Site': 'cross-site'
          },
          referrer: generateRefferer(model),
          method: 'GET',
          mode: 'cors'
        }
      )

      if (!response.ok) {
        return `{"http": "${response.status}"}`
      }

      const data = await response.json()
      return data
    } catch (error) {
      return `{"error": "${error}"}`
    }
  }

  async getObjectRelations(model: string, slug_id: string): Promise<string> {
    try {
      const response = await fetch(
        `https://${this.config.baseUrl}/api/${model}/${slug_id}/relations`,
        {
          credentials: 'include',
          headers: {
            'User-Agent': this.config.userAgent,
            Accept: '*/*',
            'Accept-Language': 'ru-RU,en-US,en',
            'Site-Id': generateSiteId(model),
            'Content-Type': 'application/json',
            'Client-Time-Zone': Intl.DateTimeFormat().resolvedOptions().timeZone,
            'Sec-Fetch-Dest': 'empty',
            'Sec-Fetch-Mode': 'cors',
            'Sec-Fetch-Site': 'cross-site'
          },
          referrer: generateRefferer(model),
          method: 'GET',
          mode: 'cors'
        }
      )

      if (!response.ok) {
        return `{"http": "${response.status}"}`
      }

      const data = await response.json()
      return data
    } catch (error) {
      return `{"error": "${error}"}`
    }
  }

  async getObjectSimilar(model: string, slug_id: string): Promise<string> {
    try {
      const response = await fetch(
        `https://${this.config.baseUrl}/api/${model}/${slug_id}/similar`,
        {
          credentials: 'include',
          headers: {
            'User-Agent': this.config.userAgent,
            Accept: '*/*',
            'Accept-Language': 'ru-RU,en-US,en',
            'Site-Id': generateSiteId(model),
            'Content-Type': 'application/json',
            'Client-Time-Zone': Intl.DateTimeFormat().resolvedOptions().timeZone,
            'Sec-Fetch-Dest': 'empty',
            'Sec-Fetch-Mode': 'cors',
            'Sec-Fetch-Site': 'cross-site'
          },
          referrer: generateRefferer(model),
          method: 'GET',
          mode: 'cors'
        }
      )

      if (!response.ok) {
        return `{"http": "${response.status}"}`
      }

      const data = await response.json()
      return data
    } catch (error) {
      return `{"error": "${error}"}`
    }
  }

  async getEpisodes(model: string = 'anime', slug_id: string): Promise<string> {
    const params = [`anime_id=${encodeURIComponent(slug_id)}`].join('&')
    try {
      const response = await fetch(
        `https://${this.config.baseUrl}/api/episodes?${params.toString()}`,
        {
          credentials: 'include',
          headers: {
            'User-Agent': this.config.userAgent,
            Accept: '*/*',
            'Accept-Language': 'ru-RU,en-US,en',
            'Site-Id': generateSiteId(model),
            'Content-Type': 'application/json',
            'Client-Time-Zone': Intl.DateTimeFormat().resolvedOptions().timeZone,
            'Sec-Fetch-Dest': 'empty',
            'Sec-Fetch-Mode': 'cors',
            'Sec-Fetch-Site': 'cross-site'
          },
          referrer: generateRefferer(model),
          method: 'GET',
          mode: 'cors'
        }
      )

      if (!response.ok) {
        return `{"http": "${response.status}"}`
      }

      const data = await response.json()
      return data
    } catch (error) {
      return `{"error": "${error}"}`
    }
  }

  async getEpisode(model: string = 'anime', episode: number): Promise<string> {
    try {
      const response = await fetch(`https://${this.config.baseUrl}/api/episodes/${episode}`, {
        credentials: 'include',
        headers: {
          'User-Agent': this.config.userAgent,
          Accept: '*/*',
          'Accept-Language': 'ru-RU,en-US,en',
          'Site-Id': generateSiteId(model),
          'Content-Type': 'application/json',
          'Client-Time-Zone': Intl.DateTimeFormat().resolvedOptions().timeZone,
          'Sec-Fetch-Dest': 'empty',
          'Sec-Fetch-Mode': 'cors',
          'Sec-Fetch-Site': 'cross-site'
        },
        referrer: generateRefferer(model),
        method: 'GET',
        mode: 'cors'
      })

      if (!response.ok) {
        return `{"http": "${response.status}"}`
      }

      const data = await response.json()
      return data
    } catch (error) {
      return `{"error": "${error}"}`
    }
  }
}

File: src-tauri/src/api/object.rs
use crate::api::config::Config;
use crate::api::utils::generate_fields::generate_fields;
use crate::api::utils::generate_request::generate_request;
use crate::api::utils::generate_url::generate_url;
use crate::api::utils::generate_user_agent::generate_user_agent;

#[derive(Clone)]
pub struct ObjectManager {
    config: Config,
    client: reqwest::Client,
}

impl ObjectManager {
    pub fn new(config: &Config, client: reqwest::Client) -> Self {
        Self {
            config: config.clone(),
            client,
        }
    }

    /// Получает обьект, его информацию и так дальше, использует слаг айди и модель.
    pub async fn get_object(
        &self,
        model: String,
        slug_id: String,
    ) -> Result<String, reqwest::Error> {
        let fields = generate_fields(&model);

        let url = generate_url(
            &self.config.base_url,
            "/{model}/{id}",
            &[("model", &model), ("id", &slug_id)],
            None,
            Some(fields.iter().map(AsRef::as_ref).collect()),
        );

        generate_request(&self.client, &url, &model, &generate_user_agent()).await
    }

    /// Получает родственников обьекта (связаные тайтлы), использует слаг айди и модель.
    pub async fn get_object_relations(
        &self,
        model: String,
        slug_id: String,
    ) -> Result<String, reqwest::Error> {
        let url = generate_url(
            &self.config.base_url,
            "/{model}/{id}/relations",
            &[("model", &model), ("id", &slug_id)],
            None,
            None,
        );

        generate_request(&self.client, &url, &model, &generate_user_agent()).await
    }

    /// Получает похожих обьекта (тайтлы что похожы по тегам, любителям, смыслу и всякому такому), использует слаг айди и модель.
    pub async fn get_object_similar(
        &self,
        model: String,
        slug_id: String,
    ) -> Result<String, reqwest::Error> {
        let url = generate_url(
            &self.config.base_url,
            "/{model}/{id}/similar",
            &[("model", &model), ("id", &slug_id)],
            None,
            None,
        );

        generate_request(&self.client, &url, &model, &generate_user_agent()).await
    }

    /// Получает епизоды обьекта, использует слаг айди и модель, скорее всего аниме онли.
    pub async fn get_object_episodes(
        &self,
        model: String,
        slug_id: String,
    ) -> Result<String, reqwest::Error> {
        let url = generate_url(
            &self.config.base_url,
            "/episodes/{id}",
            &[("id", &slug_id)],
            None,
            None,
        );

        generate_request(&self.client, &url, &model, &generate_user_agent()).await
    }
    /// Получает епизод обьекта (сам список видеопотоков), использует епизоде айди и модель, скорее всего аниме онли.
    pub async fn get_object_episode(
        &self,
        model: String,
        episode: String,
    ) -> Result<String, reqwest::Error> {
        let url = generate_url(
            &self.config.base_url,
            "/episodes",
            &[],
            Some(vec![("anime_id", &episode)]),
            None,
        );

        generate_request(&self.client, &url, &model, &generate_user_agent()).await
    }
}
